<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /home/max/catkin_ws/src/rocker_bogie/urdf/rocker_bogie.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="rocker_bogie_rov">
  <!-- Base Link at the Center of the Chassis -->
  <link name="base_link">
    <visual>
      <geometry>
        <sphere radius="0.01"/>
        <!-- Small sphere for visual debugging -->
      </geometry>
      <origin xyz="0 0 0"/>
      <!-- Centered inside the chassis -->
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <link name="chassis">
    <visual>
      <geometry>
        <box size="0.334 0.254 0.115"/>
      </geometry>
      <origin xyz="0 0 0"/>
      <material name="green">
        <color rgba="0 1 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.334 0.254 0.115"/>
      </geometry>
      <origin xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="15"/>
      <origin xyz="0 0 0"/>
      <!-- Set CoM to the geometric center -->
      <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.2"/>
    </inertial>
  </link>
  <link name="rocker_left">
    <visual>
      <!-- Horizontal Bar -->
      <geometry>
        <box size="0.32 0.024 0.024"/>
      </geometry>
      <origin xyz="-0.147 0 0"/>
      <material name="blue">
        <color rgba="0 1 1 1"/>
      </material>
    </visual>
    <visual>
      <!-- Vertical Bar -->
      <geometry>
        <box size="0.024 0.024 0.186"/>
      </geometry>
      <origin xyz="-0.294 0 -0.1"/>
      <material name="blue">
        <color rgba="0 1 1 1"/>
      </material>
    </visual>
    <visual>
      <!-- Rotating Bar -->
      <geometry>
        <box size="0.024 0.024 0.12"/>
      </geometry>
      <origin rpy="0 -0.61 0" xyz="0.055 0 -0.055"/>
      <material name="blue">
        <color rgba="0 1 1 1"/>
      </material>
    </visual>
    <collision>
      <!-- Horizontal Bar Collision -->
      <geometry>
        <box size="0.32 0.024 0.024"/>
      </geometry>
      <origin xyz="-0.147 0 0"/>
    </collision>
    <collision>
      <!-- Vertical Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.186"/>
      </geometry>
      <origin xyz="-0.294 0 -0.1"/>
    </collision>
    <collision>
      <!-- Rotating Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.12"/>
      </geometry>
      <origin rpy="0 -0.61 0" xyz="0.055 0 -0.055"/>
    </collision>
    <inertial>
      <!-- Adjusted mass to match the size -->
      <mass value="3.0"/>
      <!-- CoM aligned to the center of the rocker arm -->
      <origin xyz="-0.15 0 0"/>
      <!-- Adjusted inertia tensor values -->
      <inertia ixx="0.005" ixy="0" ixz="0" iyy="0.006" iyz="0" izz="0.007"/>
    </inertial>
  </link>
  <link name="rocker_right">
    <visual>
      <!-- Horizontal Bar -->
      <geometry>
        <box size="0.32 0.024 0.024"/>
      </geometry>
      <origin xyz="-0.147 0 0"/>
      <material name="blue">
        <color rgba="0 1 1 1"/>
      </material>
    </visual>
    <visual>
      <!-- Vertical Bar -->
      <geometry>
        <box size="0.024 0.024 0.186"/>
      </geometry>
      <origin xyz="-0.294 0 -0.1"/>
      <material name="blue">
        <color rgba="0 1 1 1"/>
      </material>
    </visual>
    <visual>
      <!-- Rotating Bar -->
      <geometry>
        <box size="0.024 0.024 0.12"/>
      </geometry>
      <origin rpy="0 -0.61 0" xyz="0.055 0 -0.055"/>
      <material name="blue">
        <color rgba="0 1 1 1"/>
      </material>
    </visual>
    <collision>
      <!-- Horizontal Bar Collision -->
      <geometry>
        <box size="0.32 0.024 0.024"/>
      </geometry>
      <origin xyz="-0.147 0 0"/>
    </collision>
    <collision>
      <!-- Vertical Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.186"/>
      </geometry>
      <origin xyz="-0.294 0 -0.1"/>
    </collision>
    <collision>
      <!-- Rotating Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.12"/>
      </geometry>
      <origin rpy="0 -0.61 0" xyz="0.055 0 -0.055"/>
    </collision>
    <inertial>
      <!-- Adjusted mass to match the size -->
      <mass value="3.0"/>
      <!-- CoM aligned to the center of the rocker arm -->
      <origin xyz="-0.15 0 0"/>
      <!-- Adjusted inertia tensor values -->
      <inertia ixx="0.005" ixy="0" ixz="0" iyy="0.006" iyz="0" izz="0.007"/>
    </inertial>
  </link>
  <link name="bogie_left">
    <visual>
      <!-- Horizontal Bar -->
      <geometry>
        <box size="0.07 0.024 0.024"/>
      </geometry>
      <origin xyz="0.177 0 0.1"/>
      <material name="orange">
        <color rgba="1 0.5 0 1"/>
      </material>
    </visual>
    <visual>
      <!-- Front Vertical Bar -->
      <geometry>
        <box size="0.024 0.024 0.162"/>
      </geometry>
      <origin xyz="0.2 0 0.01"/>
      <material name="orange">
        <color rgba="1 0.5 0 1"/>
      </material>
    </visual>
    <visual>
      <!-- Back Rotating Bar -->
      <geometry>
        <box size="0.024 0.024 0.323"/>
      </geometry>
      <origin rpy="0 0.96 0" xyz="0.01 0 0.01"/>
      <material name="orange">
        <color rgba="1 0.5 0 1"/>
      </material>
    </visual>
    <collision>
      <!-- Horizontal Bar Collision -->
      <geometry>
        <box size="0.07 0.024 0.024"/>
      </geometry>
      <origin xyz="0.177 0 0.1"/>
    </collision>
    <collision>
      <!-- Front Vertical Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.162"/>
      </geometry>
      <origin xyz="0.2 0 0.01"/>
    </collision>
    <collision>
      <!-- Back Rotating Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.323"/>
      </geometry>
      <origin rpy="0 0.96 0" xyz="0.01 0 0.01"/>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.004" ixy="0" ixz="0" iyy="0.004" iyz="0" izz="0.008"/>
    </inertial>
  </link>
  <link name="bogie_right">
    <visual>
      <!-- Horizontal Bar -->
      <geometry>
        <box size="0.07 0.024 0.024"/>
      </geometry>
      <origin xyz="0.177 0 0.1"/>
      <material name="orange">
        <color rgba="1 0.5 0 1"/>
      </material>
    </visual>
    <visual>
      <!-- Front Vertical Bar -->
      <geometry>
        <box size="0.024 0.024 0.162"/>
      </geometry>
      <origin xyz="0.2 0 0.01"/>
      <material name="orange">
        <color rgba="1 0.5 0 1"/>
      </material>
    </visual>
    <visual>
      <!-- Back Rotating Bar -->
      <geometry>
        <box size="0.024 0.024 0.323"/>
      </geometry>
      <origin rpy="0 0.96 0" xyz="0.01 0 0.01"/>
      <material name="orange">
        <color rgba="1 0.5 0 1"/>
      </material>
    </visual>
    <collision>
      <!-- Horizontal Bar Collision -->
      <geometry>
        <box size="0.07 0.024 0.024"/>
      </geometry>
      <origin xyz="0.177 0 0.1"/>
    </collision>
    <collision>
      <!-- Front Vertical Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.162"/>
      </geometry>
      <origin xyz="0.2 0 0.01"/>
    </collision>
    <collision>
      <!-- Back Rotating Bar Collision -->
      <geometry>
        <box size="0.024 0.024 0.323"/>
      </geometry>
      <origin rpy="0 0.96 0" xyz="0.01 0 0.01"/>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.004" ixy="0" ixz="0" iyy="0.004" iyz="0" izz="0.008"/>
    </inertial>
  </link>
  <link name="cylinder_connector_left">
    <visual>
      <geometry>
        <cylinder length="0.03" radius="0.0075"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0.145 0"/>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.028" radius="0.0075"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0.145 0"/>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <origin xyz="0 0.145 0"/>
      <!-- CoM aligned with connector -->
      <inertia ixx="0.00005" ixy="0" ixz="0" iyy="0.00005" iyz="0" izz="0.0001"/>
    </inertial>
  </link>
  <link name="cylinder_connector_right">
    <visual>
      <geometry>
        <cylinder length="0.03" radius="0.0075"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 -0.145 0"/>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.028" radius="0.0075"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 -0.145 0"/>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <origin xyz="0 -0.145 0"/>
      <!-- CoM aligned with connector -->
      <inertia ixx="0.00005" ixy="0" ixz="0" iyy="0.00005" iyz="0" izz="0.0001"/>
    </inertial>
  </link>
  <joint name="base_to_chassis" type="fixed">
    <parent link="base_link"/>
    <child link="chassis"/>
    <origin xyz="0 0 0"/>
  </joint>
  <joint name="joint_chassis_to_cylinder_left" type="fixed">
    <parent link="chassis"/>
    <child link="cylinder_connector_left"/>
    <origin xyz="0 0 0"/>
  </joint>
  <joint name="joint_chassis_to_cylinder_right" type="fixed">
    <parent link="chassis"/>
    <child link="cylinder_connector_right"/>
    <origin xyz="0 0 0"/>
  </joint>
  <joint name="joint_cylinder_left_to_rocker_left" type="revolute">
    <parent link="cylinder_connector_left"/>
    <child link="rocker_left"/>
    <origin xyz="0 0.165 0"/>
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="-0.26" upper="0.26" velocity="100"/>
    <dynamics damping="0.5" spring_stiffness="50"/>
    <gazebo>
      <spring_reference>0.0</spring_reference>
      <!-- Neutral position -->
      <spring_stiffness>50.0</spring_stiffness>
      <!-- Spring constant -->
      <damping>0.5</damping>
      <!-- Damping coefficient -->
    </gazebo>
  </joint>
  <joint name="joint_cylinder_left_to_rocker_right" type="revolute">
    <parent link="cylinder_connector_right"/>
    <child link="rocker_right"/>
    <origin xyz="0 -0.165 0"/>
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="-0.26" upper="0.26" velocity="100"/>
    <dynamics damping="0.5" spring_stiffness="50"/>
    <gazebo>
      <spring_reference>0.0</spring_reference>
      <!-- Neutral position -->
      <spring_stiffness>50.0</spring_stiffness>
      <!-- Spring constant -->
      <damping>0.5</damping>
      <!-- Damping coefficient -->
    </gazebo>
  </joint>
  <joint name="joint_rocker_to_bogie_left" type="revolute">
    <parent link="rocker_left"/>
    <child link="bogie_left"/>
    <origin xyz="0.1 0 -0.1"/>
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="-0.26" upper="0.26" velocity="100"/>
    <dynamics damping="0.5" spring_stiffness="50"/>
    <gazebo>
      <spring_reference>0.0</spring_reference>
      <!-- Neutral position -->
      <spring_stiffness>50.0</spring_stiffness>
      <!-- Spring constant -->
      <damping>0.5</damping>
      <!-- Damping coefficient -->
    </gazebo>
  </joint>
  <joint name="joint_rocker_to_bogie_right" type="revolute">
    <parent link="rocker_right"/>
    <child link="bogie_right"/>
    <origin xyz="0.1 0 -0.1"/>
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="-0.26" upper="0.26" velocity="100"/>
    <dynamics damping="0.5" spring_stiffness="50"/>
    <gazebo>
      <spring_reference>0.0</spring_reference>
      <!-- Neutral position -->
      <spring_stiffness>50.0</spring_stiffness>
      <!-- Spring constant -->
      <damping>0.5</damping>
      <!-- Damping coefficient -->
    </gazebo>
  </joint>
  <link name="wheel_front_left">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <material name="gray">
        <color rgba="0.3 0.3 0.3 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0005" ixy="0" ixz="0" iyy="0.0005" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  <link name="wheel_middle_left">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <material name="gray">
        <color rgba="0.3 0.3 0.3 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0005" ixy="0" ixz="0" iyy="0.0005" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  <link name="wheel_back_left">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <material name="gray">
        <color rgba="0.3 0.3 0.3 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0005" ixy="0" ixz="0" iyy="0.0005" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  <link name="wheel_front_right">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <material name="gray">
        <color rgba="0.3 0.3 0.3 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0005" ixy="0" ixz="0" iyy="0.0005" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  <link name="wheel_middle_right">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <material name="gray">
        <color rgba="0.3 0.3 0.3 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0005" ixy="0" ixz="0" iyy="0.0005" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  <link name="wheel_back_right">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <material name="gray">
        <color rgba="0.3 0.3 0.3 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0005" ixy="0" ixz="0" iyy="0.0005" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  <joint name="joint_wheel_front_left" type="continuous">
    <parent link="bogie_left"/>
    <!-- Attach the wheel to the correct parent link (e.g., bogie or rocker) -->
    <child link="wheel_front_left"/>
    <origin xyz="0.2 0.03 -0.07"/>
    <axis xyz="0 1 0"/>
    <!-- Rotate around y-axis -->
  </joint>
  <joint name="joint_wheel_middle_left" type="continuous">
    <parent link="bogie_left"/>
    <!-- Attach the wheel to the correct parent link (e.g., bogie or rocker) -->
    <child link="wheel_middle_left"/>
    <origin xyz="-0.12 0.03 -0.07"/>
    <axis xyz="0 1 0"/>
    <!-- Rotate around y-axis -->
  </joint>
  <joint name="joint_wheel_back_left" type="continuous">
    <parent link="rocker_left"/>
    <!-- Attach the wheel to the correct parent link (e.g., bogie or rocker) -->
    <child link="wheel_back_left"/>
    <origin xyz="-0.3 0.03 -0.17"/>
    <axis xyz="0 1 0"/>
    <!-- Rotate around y-axis -->
  </joint>
  <joint name="joint_wheel_front_right" type="continuous">
    <parent link="bogie_right"/>
    <!-- Attach the wheel to the correct parent link (e.g., bogie or rocker) -->
    <child link="wheel_front_right"/>
    <origin xyz="0.2 -0.03 -0.07"/>
    <axis xyz="0 1 0"/>
    <!-- Rotate around y-axis -->
  </joint>
  <joint name="joint_wheel_middle_right" type="continuous">
    <parent link="bogie_right"/>
    <!-- Attach the wheel to the correct parent link (e.g., bogie or rocker) -->
    <child link="wheel_middle_right"/>
    <origin xyz="-0.12 -0.03 -0.07"/>
    <axis xyz="0 1 0"/>
    <!-- Rotate around y-axis -->
  </joint>
  <joint name="joint_wheel_back_right" type="continuous">
    <parent link="rocker_right"/>
    <!-- Attach the wheel to the correct parent link (e.g., bogie or rocker) -->
    <child link="wheel_back_right"/>
    <origin xyz="-0.3 -0.03 -0.17"/>
    <axis xyz="0 1 0"/>
    <!-- Rotate around y-axis -->
  </joint>
  <gazebo>
    <plugin filename="libskid_steering_plugin.so" name="skid_steering_plugin"/>
  </gazebo>
  <!-- 20/10  -->
</robot>
